MAINT update dependencies to 2023 releases + numba_dpex 0.19.0

fcharras opened the pull request:
Last remaining TODO: euclidean distances unit test failing on GPU.

fcharras commented:
I think the PR is in a good state to start review.

What's in it:

- [x] bumps all dependencies to most recent versions in the Dockerfile
- [x] fixes issues related to kernels not running if local memory overflows after oneapi 2023 releases https://github.com/IntelPython/numba-dpex/issues/865
- [x] fixes issues related to spirv versioning in numba_dpex after oneapi 2023 releases https://github.com/IntelPython/numba-dpex/issues/865
- [x] workaround for bug in numba_dpex 0.19.0 (probably related to bump to latest numba version) https://github.com/IntelPython/numba-dpex/issues/867

Also, according to https://github.com/IntelPython/numba-dpex/issues/778 it's not necessary to install the whole oneapi basekit to only use the `llvm_spirv` binary. This PR improves the dockerfile in this regard.

Remaining TODOs;
- [ ] fix the last remaining failing test (locally), caused by weird numerical errors in euclidean distances
- [ ] test and push separately the new dockerfile to trigger rebuild of the up-to-date image and let the CI be green in this PR
- [ ] review & merge the PR

jjerphan commented on docker/Dockerfile:
```diff
@@ -155,10 +135,13 @@ RUN mkdir -p $ONEAPI_INSTALL_DIR $ONEAPI_CACHE_DIR $ONEAPI_DOWNLOAD_DIR $ONEAPI_
-    && rm -Rf $TMPDIR/*
+    && rm -Rf $TMPDIR/* \
+    && mkdir -p $LLVM_SPIRV_INSTALL_DIR \
+    && mv $ONEAPI_INSTALL_DIR/compiler/latest/linux/bin-llvm/llvm-spirv $LLVM_SPIRV_INSTALL_DIR \
+    && ln -s $LLVM_SPIRV_INSTALL_DIR/llvm-spirv $ONEAPI_INSTALL_DIR/compiler/latest/linux/bin-llvm
```
I think the rational of separating `LLVM_SPIRV_INSTALL_DIR` and `ONEAPI_INSTALL_DIR` must be documented.

fcharras replied:
I changed my mind and removed the changes in this block and only extract the `llvm_spirv` binary at the very end (see https://github.com/soda-inria/sklearn-numba-dpex/pull/83).

Adding a comment to explain why.

jjerphan commented on sklearn_numba_dpex/common/_utils.py:
```diff
@@ -1,23 +1,148 @@
+
+
+def _force_reload_numba_dpex_with_patches(with_spirv_fix=True):
+    """This function hacks `numba_dpex` init to work around issues after
+    `dpctl>=0.14.1dev1` and `numba_dpex>=0.19.0` bumps. It will be
+    reverted when the official fixes are out.
+    """
```
Should we link to an issue?

fcharras replied:
Added a link to the relevant PR

jjerphan commented on sklearn_numba_dpex/common/_utils.py:
```diff
@@ -1,23 +1,148 @@
+        return _mock_device()
+
+    try:
+        # A better fix for this is already available in the development tree of
+        # `numba_dpex` but it's not released yet.
```
Can you add a permalink here, please? :angel:

jjerphan commented on sklearn_numba_dpex/common/kernels.py:
```diff
@@ -379,6 +385,20 @@ def add_first_cols(from_array, to_array, to_col):
+    if work_group_size == input_work_group_size:
+        check_power_of_2(work_group_size)
+    else:
+        work_group_size = 2 ** (math.floor(math.log2(work_group_size)))
```
```suggestion
        # Round to the maximum smaller power of two 
        work_group_size = 2 ** (math.floor(math.log2(work_group_size)))
```

jjerphan commented on sklearn_numba_dpex/common/kernels.py:
```diff
@@ -433,22 +453,32 @@ def partial_sum_reduction(
+    if work_group_size == input_work_group_size:
+        check_power_of_2(work_group_size)
+    else:
+        work_group_size = 2 ** (math.floor(math.log2(work_group_size)))
```
```suggestion
        # Round to the maximum smaller power of two 
        work_group_size = 2 ** (math.floor(math.log2(work_group_size)))
```

jjerphan commented on sklearn_numba_dpex/kmeans/kernels/compute_labels.py:
```diff
@@ -4,6 +4,8 @@
 import numba_dpex as dpex
 import numpy as np
 
+from sklearn_numba_dpex.common._utils import _check_max_work_group_size
```
Can we use relative import paths?

This comment applies in other places.

fcharras replied:
We already have import path when the related files are not too far away in the file tree.

Here it would be `from ...common._utils` isn't explicit path better than `...` ?

jjerphan commented on sklearn_numba_dpex/tests/test_patches.py:
```diff
@@ -0,0 +1,84 @@
+
+from sklearn_numba_dpex.common._utils import _force_reload_numba_dpex_with_patches
+
+
```
I like those tests' rationale. :+1:

jjerphan commented on sklearn_numba_dpex/common/kernels.py:
```diff
@@ -295,15 +301,16 @@ def _make_partial_sum_reduction_2d_axis1_kernel(
 
-    fused_unary_func = dpex.func(fused_unary_func)
+    else:
+        fused_unary_func = dpex.func(fused_unary_func())
```
Is it normal to call `fused_unary_func`, here? Shouldn't we just pass is as an argument? If that's a workaround, can you document it (as temporarily)?
```suggestion
        fused_unary_func = dpex.func(fused_unary_func)
```

fcharras replied:
I moved this trick from there to the driver file, it was easier to explain in a comment there.

jjerphan commented on sklearn_numba_dpex/common/kernels.py:
```diff
@@ -135,7 +142,7 @@ def make_broadcast_ops_1d_2d_axis1_kernel(size0, size1, ops, work_group_size):
 
     global_size = math.ceil(size1 / work_group_size) * work_group_size
-    ops = dpex.func(ops)
+    ops = dpex.func(ops())
```
Similarly, is it normal to call `ops`, here? Shouldn't we just pass is as an argument? If that's a workaround, can you document it (as temporarily)?

jjerphan commented on sklearn_numba_dpex/common/_utils.py:
```diff
@@ -1,23 +1,148 @@
+):
+    """For CPU devices, the value `device.max_work_group_size` seems to always be
+    surprisingly large, up to several order of magnitude higher than the number of
+    threads (for instance, having 8 threads and a `max_work_group_size` equal to
+    8192). It means that, for CPUs, a work group schedules big batches of tasks per
+    thread, and that only one work group will be executed at a time. Kernels that
+    allocate an amount of local memory (i.e fast access memory shared by the work
+    group) that scale with the size of the work group are at risk of overflowing the
+    size of the local memory (given by `device.local_mem_size`, typically 32kB). So
+    we need to scale down the size of the work groups with respect to
+    `device.local_mem_size` to prevent overflowing.
 
-def _plus(x, y):
-    return x + y
+    This is not an issue with GPU devices, for our kernels usually respect the rule of
+    thumb of allocating in local memory about one item per thread, which fit the GPU
+    architecture well and seems to be enough to prevent overflowing. With GPUs, max
+    possible work group sizes are smaller, such that there's no oversubscription of
+    tasks and GPUs will execute tasks of several work groups at once. For GPUs, this
+    approach at local memory allocation is by design reliable, the amount of available
+    local memory is enough to ensure that the memory needed by the running compute is
+    allocated. As a consequence, the checks that are enforced for CPUs are not needed.
 
+    NB: this function only applies an upper bound to the `work_group_size`. It might
+    still be needed to apply other requirements, such that being a multiple of
+    `sub_group_size` and/or a power of two.
+    """
```
Thanks! :100:

jjerphan reviewed (commented):
Thank you for working around limitations of the latest releases, @fcharras.

Here is a first and hopefully last review. I think we should not spend too much time perfecting those workaround if they are temporarily but make sure enough information is present to revert them once the limitations are addressed upstream.

fcharras commented on docker/Dockerfile:
```diff
@@ -60,9 +60,7 @@ ARG INTEL_SCIPY_VERSION="==1.7.3"
-# NB: `numba_dpex>=0.19` should be compatible with `numba>=0.56` but it does not seems
-# to work at the moment, see https://github.com/IntelPython/numba-dpex/issues/850
-ARG INTEL_NUMBA_VERSION="<0.56"
+ARG INTEL_NUMBA_VERSION=""
```
I think it's fine to let blank here. I gives us automatic update once a week when the image is rebuilt by the ci. We can pin if it happens to break again.

fcharras commented on pyproject.toml:
```diff
@@ -37,8 +37,8 @@ classifiers = [
 # see https://github.com/IntelPython/dpctl/issues/886
 dependencies = [
     "scikit-learn",
```
AFAIK there's no syntax for this in the dependencies metedata.

fcharras commented:
- Addressed suggestions
- I pushed the Dockerfile on main already, the image has been rebuilt, and the CI is green now
- Nice improvement: the image size is down from 10GB to 5GB :feather:

fcharras commented:
The last remaining failing test was actually a bigger issue than expected, but still related to `numba_dpex` `dpex.func` caching. I added an additional hack to `numba_dpex` internals to ensure that some `func` are not wrongly used in place of others (basically adding `id(func)` to `func.__name__` and `func.__qualname__` before submitting to compiler)

ogrisel commented:
> I added an additional hack to numba_dpex internals to ensure that some func are not wrongly used in place of others (basically adding id(func) to func.__name__ and func.__qualname__ before submitting to compiler)

Did you report this usability problem upstream already?

ogrisel commented on docker/Dockerfile:
```diff
@@ -332,6 +332,13 @@ RUN --mount=type=bind,target=$ONEAPI_INSTALL_DIR,source=$ONEAPI_INSTALL_DIR,from
+# However it lacks the `llvm-spirv` binary, that we extract from the oneapi basekit
+# and that we can keep as a standalone binary, since it does not link against any of
+# the other binaries of the oneapi basekit (see
+# https://github.com/IntelPython/numba-dpex/issues/778#issuecomment-1271617194 )
```
Thanks for the comment and reference!

fcharras commented:
> Did you report this usability problem upstream already?

 Yes links to the relevant issues and PRs are inserted in inline comments next to the hacks. The issues with caching device funcs in `0.19.0` were known already, and `numba_dpex` devs are actively working on this at the moment.

ogrisel commented on sklearn_numba_dpex/common/kernels.py:
```diff
@@ -102,7 +109,7 @@ def initialize_to_zeros(data):
 
 @lru_cache
-def make_broadcast_division_1d_2d_axis0_kernel(size0, size1, work_group_size):
+def make_broadcast_division_1d_2d_axis0_kernel(size0, size1, work_group_size, dtype):
```
Is passing the dtype requiring here as well? There is no occurrence of `dpex.func` in this function, only a `dpex.kernel`.

Or maybe we can keep it so as to get a consistent signature style for all kernel factory function, irrespective of the fact that they use device functions or not.

fcharras replied:
You're right it's a mistake, thanks for pointing it out.

ogrisel commented on sklearn_numba_dpex/common/kernels.py:
```diff
@@ -24,13 +24,20 @@
 zero_idx = np.int64(0)
 
 
+# HACK: dtype argument is passed to prevent caching as a workaround for
```
```suggestion
# HACK: dtype argument is passed to prevent sharing a device function
# instance between kernels specialized for different argument types.
# This is a workaround for:
```

ogrisel commented on sklearn_numba_dpex/kmeans/drivers.py:
```diff
@@ -6,6 +6,9 @@
 
+# HACK: temporarily, those operators are wrapped in a function that must be called
+# without arguments to get the operator.
+# See the notice where it's defined for more informations.
```
```suggestion
# See the notice where they are defined for more information.
```

ogrisel reviewed (approved):
LGTM!

Just a few comments / suggestions below.

fcharras commented:
Thank you for the reviews @jjerphan @ogrisel , I've applied all suggestions, I'll merge after the test pipeline has run.
